apis/controlplane/v1alpha3/nestedcontrolplane_types.go

```diff
@@ -0,0 +1,107 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha3
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// NestedControlPlaneSpec defines the desired state of NestedControlPlane
+type NestedControlPlaneSpec struct {
+	// ClusterNamespace allows you to specify the namespace where the control
+	// plane components are run, if you don't specify one then the namespace
+	// will be in the format of {namespace}-{vcp.uid}. This also makes up the
+	// prefixes used for all Namespaces in the cluster.
+	// e.g. {namespace}-{vcp.uid}-{vc-namespace}
+	// +optional
+	ClusterNamespace string `json:"clusterNamespace,omitempty"`
+
+	// ClusterDomain allows you to specify the domain name of the nested
+	// cluster. e.g. a pod dns will be
+	// {some-pod}.{some-namespace}.svc.{ClusterDomain}
+	// +kubebuilder:default=cluster.local
+	ClusterDomain string `json:"clusterDomain,omitempty"`
+
+	// TemplateRef allows you to specify the control plane template to create
+	// the control plane.
+	// +optional
+	TemplateRef *corev1.ObjectReference `json:"configRef,omitempty"`
+
+	// PKIExpireDays allows you to specify the period of time the nested
+	// cluster PKI, if not set the PKI will expire after 1 year.
+	// +optional
+	// +kubebuilder:default=365
+	PKIExpireDays int64 `json:"pkiExpireDays,omitempty"`
```

Comment:
I saw PKI and had some quick thoughts:
- it feels rather coarse to set the expiry at the granularity of the whole PKI, rather than specific credentials
- are you thinking about rotation yet?
- Why put this in NestedControlPlaneSpec instead of NestedPKI?

Reply:
I think we should err on the side of caution with adding fields. If we don't need to expose this as a user-configurable option yet, I think we should not do it. The above questions raise great points, and I am not sure we have enough of an understanding of the problem space to know where best to put it just yet.

> Why put this in NestedControlPlaneSpec instead of NestedPKI?

👍 - I also think this is a question we should answer more generally. What is the purpose of the NestedControlPlane vs the NestedControlPlaneTemplate resource? What defines what should go where?